edit/edit.go

```diff
@@ -461,6 +476,150 @@ func ContainsComments(expr build.Expr, str string) bool {
 	return false
 }
 
+// RemoveEmptySelectsAndConcatLists iterates the tree in order to turn
+// empty selects into empty lists and adjacent lists are concatenated
+func RemoveEmptySelectsAndConcatLists(e build.Expr) build.Expr {
+	switch e := e.(type) {
+	case *build.BinaryExpr:
+		if e.Op == "+" {
+			e.X = RemoveEmptySelectsAndConcatLists(e.X)
+			e.Y = RemoveEmptySelectsAndConcatLists(e.Y)
+
+			x, xIsList := e.X.(*build.ListExpr)
+			y, yIsList := e.Y.(*build.ListExpr)
+
+			if xIsList && yIsList {
+				return &build.ListExpr{List: append(x.List, y.List...)}
+			}
+
+			if xIsList && len(x.List) == 0 {
+				return e.Y
+			}
+
+			if yIsList && len(y.List) == 0 {
+				return e.X
+			}
+		}
+	case *build.CallExpr:
+		if x, ok := e.X.(*build.Ident); ok && x.Name == "select" {
+			dict := e.List[0].(*build.DictExpr)
+			for _, keyVal := range dict.List {
+				keyVal := keyVal.(*build.KeyValueExpr)
+				val, ok := keyVal.Value.(*build.ListExpr)
+				if !ok || len(val.List) > 0 {
+					return e
+				}
+			}
+
+			return &build.ListExpr{List: []build.Expr{}}
+		}
+	}
+
+	return e
+}
+
+// SelectListsIntersection returns the intersection of the lists of strings inside
+// the dictionary argument of the select expression given as a parameter
+func SelectListsIntersection(sel *build.CallExpr, pkg string) (intersection []build.Expr) {
+	dict := sel.List[0].(*build.DictExpr)
+	firstList := dict.List[0].(*build.KeyValueExpr).Value.(*build.ListExpr).List
```

Comment:
Same here (in the two lines above). You're assuming that 1. the select has an argument, 2. it's a dict literal, 3. the dict is non-empty, 4. its subnodes are KeyValueExpr nodes, 5. the value of the virst key-value pair is a list literal. If any of these assumptions are not correct, buildozer will fail.

Although if the 4th assumption is wrong it means there's a bug somewhere, all other cases can occur in input files (even if some of these files are not valid for Bazel).

Even if you're sure that the value must be a list, it's not necessarily a list literal, it can be a constant or an expression, the computed value of which is a list, e.g. `getParams(foo).split(";")`, where `getParams` returns a string.

Please add tests for these cases to make sure they don't crash buildozer.

Reply:
Yes, I didn't take such possibilities into account. Thanks! :D